### PR TITLE
mon: define paxos services indexes using enum and cleanups

### PR DESCRIPTION
**Pull request by changchengx**
1. remove unused function
2. define paxos services indexes using enum
3. define paxos_service with fixed array size

Signed-off-by: Changcheng Liu <changcheng.liu@aliyun.com>

## Checklist
- [ ] References tracker ticket
- [ ] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

---

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

### CODE REVIEW

**Comment by changchengx** [comment]
@tchaikov 
I'm trying to add one PAXOS Service for one developing project.
Historically, some people had extended the service type, e.g. https://github.com/ceph/ceph/commit/38db0a289f572fa40542bd96a0d9ec717ac3db68#diff-12b42de6db7cf2045db70216aab8b86bR34
This PR makes it easy to extend PAXOS service type.

**Comment by changchengx** [comment]
jenkins, test this please

**Comment by changchengx** [comment]
jenkins, test this please

**Comment by changchengx** [comment]
@tchaikov @jecluis Please help review this PR

**Comment by changchengx** [comment]
@jecluis ping

**Comment by changchengx** [comment]
@tchaikov ping

File: src/mon/mon_types.h [hunk 1]
```diff
@@ -24,16 +24,19 @@
...
-#define PAXOS_LOG        2
-#define PAXOS_MONMAP     3
-#define PAXOS_AUTH       4
-#define PAXOS_MGR        5
-#define PAXOS_MGRSTAT    6
-#define PAXOS_HEALTH     7
-#define PAXOS_CONFIG     8
-#define PAXOS_NUM        9
+enum PAXOS_SVC_TYPE : int {
+  PAXOS_MDSMAP,
```
Comment (tchaikov) [inline src/mon/mon_types.h:28]: please don't use ALL_CAPTITAL_CHARACTERS as the name of the enum type, actually it is optional. i wonder if it'd be better to just to skip it.

might want to add a comment to document that these enum are used as the indexes of `Monitor::paxos_service`.

BTW, we could probably define it as a `std::array<>` as its size is known at compile-time.

File: src/mon/mon_types.h [hunk 2]
```diff
@@ -37,9 +37,9 @@ enum PAXOS_SVC_TYPE : int {
   PAXOS_NUM
 };
 
-
-inline const char *get_paxos_name(int p) {
-  switch (p) {
+template <const enum PAXOS_SVC_TYPE paxos_svc_type = PAXOS_NUM>
+static constexpr std::string get_paxos_name() {
```
Comment (tchaikov) [inline src/mon/mon_types.h:41]: i'd suggest just drop this helper. no need to improve a dead function.

File: src/mon/HealthMonitor.cc [hunk 3]
```diff
@@ -67,8 +67,8 @@ static ostream& _prefix(std::ostream *_dout, const Monitor *mon,
 		<< "(" << mon->get_state_name() << ").health ";
 }
 
-HealthMonitor::HealthMonitor(Monitor *m, Paxos *p, const string& service_name)
-  : PaxosService(m, p, service_name) {
+HealthMonitor::HealthMonitor(Monitor *m, Paxos *p)
+  : PaxosService(m, p, get_paxos_name<PAXOS_HEALTH>()) {
```
Comment (tchaikov) [inline src/mon/HealthMonitor.cc:71]: i am not in favor of this change. as it trades readability for flexibility which is not necessary here for obvious reasons.

File: src/mon/mon_types.h (hunk 2 above)
Comment (changchengx) [reply src/mon/mon_types.h:41]: @tchaikov 
I want to use this function to stringify the enum type member after defining the enum type member.
When extending the paxos service, it's easier to add the enum type member and set the paxos service name at nearby places.

Comment (tchaikov) [reply src/mon/mon_types.h:41]: i think i'd be better to keep the name of service closer to where the service is defined instead of centralizing them.

Comment (changchengx) [reply src/mon/mon_types.h:41]: Got it.
I'll remove get_paxos_name function here.

File: src/mon/HealthMonitor.cc (hunk 3 above)
Comment (changchengx) [reply src/mon/HealthMonitor.cc:71]: I'll abandon this commit since I decided to remove get_paxos_name function.

File: src/mon/mon_types.h (hunk 1 above)
Comment (changchengx) [reply src/mon/mon_types.h:28]: 1. I'll add comment here
2. I'll try to change the type of ```Monitor::paxos_service``` from std::vector to std::array

**Review by jecluis — CHANGES_REQUESTED** [review]
Aside from capturing the indexes in an enum or array, I don't see a particular reason for the other changes. And I agree with @tchaikov when it's said that we are trading readability for some flexibility; IMO this flexibility is too small to justify the tradeoff.

I'd suggest dropping those commits and go forth with the enum/whatever change.

File: src/mon/mon_types.h [hunk 4]
```diff
@@ -24,16 +24,20 @@
...
-#define PAXOS_LOG        2
-#define PAXOS_MONMAP     3
-#define PAXOS_AUTH       4
-#define PAXOS_MGR        5
-#define PAXOS_MGRSTAT    6
-#define PAXOS_HEALTH     7
-#define PAXOS_CONFIG     8
-#define PAXOS_NUM        9
+// use as paxos_service index
+enum paxos_svc_type : int {
```
Comment (tchaikov) [inline src/mon/mon_types.h:28]:
```suggestion
enum {
```

Comment (tchaikov) [reply src/mon/mon_types.h:28]: nit, the title of "enumerate paxos service type for scability" does not look quite right. IMHO, better off put something like

> mon/mon_types.h: define paxos services indexes using enum
>
> simpler this way

File: src/mon/Monitor.h [hunk 5]
```diff
@@ -637,7 +637,7 @@ class Monitor : public Dispatcher,
   /**
    * Vector holding the Services serviced by this Monitor.
    */
-  std::vector<std::unique_ptr<PaxosService>> paxos_service;
```
Comment (tchaikov) [inline src/mon/Monitor.h:640]: could you `#include <array>` to ensure that the header is self-contained?

File: src/mon/Monitor.cc [hunk 6]
```diff
@@ -203,7 +203,6 @@ Monitor::Monitor(CephContext* cct_, string nm, MonitorDBStore *s,
   timecheck_rounds_since_clean(0),
   timecheck_event(NULL),
 
-  paxos_service(PAXOS_NUM),
```
Comment (tchaikov) [inline src/mon/Monitor.cc:206]: better off change the prefix of the title to "mon/Monitor:". this change does not touch the paxos implementation.

File: src/mon/mon_types.h [hunk 7]
```diff
@@ -38,22 +38,6 @@ enum paxos_svc_type : int {
   PAXOS_NUM
 };
 
-
```
Comment (tchaikov) [inline src/mon/mon_types.h:37]: might want to change the prefix of the title to "mon/mon_types.h".

File: src/mon/Monitor.cc (hunk 6 above)
Comment (changchengx) [reply src/mon/Monitor.cc:206]: Done

File: src/mon/mon_types.h (hunk 7 above)
Comment (changchengx) [reply src/mon/mon_types.h:37]: Done

File: src/mon/Monitor.h (hunk 5 above)
Comment (changchengx) [reply src/mon/Monitor.h:640]: Done

File: src/mon/mon_types.h (hunk 4 above)
Comment (changchengx) [reply src/mon/mon_types.h:28]: Done

File: src/mon/mon_types.h [hunk 8]
```diff
@@ -24,16 +24,19 @@
...
-#define PAXOS_LOG        2
-#define PAXOS_MONMAP     3
-#define PAXOS_AUTH       4
-#define PAXOS_MGR        5
-#define PAXOS_MGRSTAT    6
-#define PAXOS_HEALTH     7
-#define PAXOS_CONFIG     8
-#define PAXOS_NUM        9
+// use as paxos_service index
+enum {
```
Comment (tchaikov) [inline src/mon/mon_types.h:28]: please note, `enum` does not define a "scoped enum type", `enum class` or `enum struct` does.

**Review by tchaikov — APPROVED** [review]

**Comment by tchaikov** [comment]
@changchengx also i changed the title of this PR from

> mon/paxos: make it easy to scale paxos service

to 

> mon: define paxos services indexes using enum and cleanups

as the former explains the original intention of this PR but it fails to reflects the exact change made in this changeset.

**Review by jecluis — APPROVED** [review]
looks okay to me.

**Comment by tchaikov** [comment]
- https://pulpito.ceph.com/kchai-2020-10-17_08:59:57-rados-wip-kefu-testing-2020-10-17-1516-distro-basic-smithi/
- https://pulpito.ceph.com/kchai-2020-10-18_16:30:00-rados-wip-kefu-testing-2020-10-18-2239-distro-basic-smithi/
- https://pulpito.ceph.com/kchai-2020-10-19_04:31:12-rados-wip-kefu-testing-2020-10-18-2239-distro-basic-smithi/
- https://pulpito.ceph.com/kchai-2020-10-19_06:06:20-rados-wip-kefu-testing-2020-10-18-2239-distro-basic-smithi

failures tracked by

- https://tracker.ceph.com/issues/47024

**Pull request merged**